cmd/source_profile.go

```diff
@@ -0,0 +1,274 @@
+			default:
+				return "", fmt.Errorf("invalid source database %v for source-profile type file", strings.ToLower(source))
+			}
+		}
+	case SourceProfileTypeConnection:
+		{
+			switch strings.ToLower(source) {
+			case "mysql":
+				return "mysql", nil
+			case "postgresql", "postgres", "pg":
+				return "postgres", nil
+			case "dynamodb":
+				return "dynamodb", nil
+			default:
+				return "", fmt.Errorf("invalid source database %v for source-profile direct connect", strings.ToLower(source))
+			}
+		}
+	case SourceProfileTypeConfig:
+		return "", fmt.Errorf("specifying source-profile using config not implemented")
+	default:
+		return "", fmt.Errorf("invalid source-profile, could not infer type")
+	}
+}
+
+// Source profile is passed as a list of key value pairs on the command line.
+// Following 3 formats are supported as valid source profiles.
+//
+// Format 1. Specify file path and file format.
+// File path can be a local file path or a gcs file path. Support for more file
+// path types can be added in future.
+// File format can be "dump" e.g., when specifying a mysqldump or pgdump etc.
+// Support for more formats e.g., "csv", "avro" etc can be added in future.
+//
+// Example: -source-profile="file=/tmp/abc, format=dump"
+// Example: -source-profile="file=gcs://bucket_name/cart.txt, format=dump"
+//
+// Format 2. Specify source connection parameters. If none specified, then read
+// from envrironment variables.
+//
+// Format 3. Specify a config file that specifies source connection profile.
+//
+func NewSourceProfile(str string, source string) (SourceProfile, error) {
+	params := make(map[string]string)
+
+	if len(str) > 0 {
+		kvs := strings.Split(str, ",")
+		for _, kv := range kvs {
+			s := strings.Split(strings.TrimSpace(kv), "=")
+			if len(s) != 2 {
+				return SourceProfile{}, fmt.Errorf("invalid key, value in source profile (expected format: key1=value1): %v", kv)
+			}
+			if _, ok := params[s[0]]; ok {
+				return SourceProfile{}, fmt.Errorf("duplicate key in source profile: %v", s[0])
+			}
+			params[s[0]] = s[1]
+		}
+	}
+
+	if _, ok := params["file"]; ok || filePipedToStdin() {
+		profile := NewSourceProfileFile(params)
+		return SourceProfile{ty: SourceProfileTypeFile, file: profile}, nil
+	} else if file, ok := params["config"]; ok {
+		config := NewSourceProfileConfig(file)
+		return SourceProfile{ty: SourceProfileTypeConfig, config: config}, fmt.Errorf("source-profile type config not yet implemented")
+	} else {
+		// Assume connection profile type connection by default, since
+		// connection parameters could be specified as part of environment
+		// variables.
+		conn, err := NewSourceProfileConnection(source, params)
+		return SourceProfile{ty: SourceProfileTypeConnection, conn: conn}, err
+	}
+}
+
+var filePipedToStdin = func() bool {
+	stat, _ := os.Stdin.Stat()
+	// Data is being piped to stdin, if true. Else, stdin is from a terminal.
+	return (stat.Mode() & os.ModeCharDevice) == 0
```

Comment:
Question: this is true i.e. piped to stdin for both supported ways to ingest a dump? Using the | and <  operators in the command?

Reply:
Yes.